### ISSUE TITLE
Encrypted raw entries cannot be created

### ISSUE THREAD

**Issue by vszakats** [nmoinvaz/minizip#284]
https://github.com/nmoinvaz/minizip/commit/ee614ff3ce14290eae66da534b796571ebb433b5#diff-eb9d271ce0492183894ba6ed179b2b0a

The above hunk, more specifically the added `|| password != NULL` expression is causing a regression where corrupted .zip entry is created after copying an encrypted raw entry from the old .zip to the new one. This happens regardless of a password value passed. Password was passed before as a zero length string, to steer the logic to the expected behaviour, but changing that to NULL doesn't fix it after this patch.

This is so because `mz_zip_entry_write_open()` would only set the `MZ_ZIB_FLAG_ENCRYPTED` flag if `szPassword != NULL`, but the above logic requires `szPassword == NULL` to create valid output.

I tried to work this around by manually setting the encrypted bit in the flag passed to `mz_zip_entry_write_open()`, but that doesn't work either because this function will reset it according to the `szPassword` value, meaning it will force clear it in this case.

I'm not sure what would be a correct fix, but as a temporary solution, commenting out `|| password != NULL` resolves the problem.

**Comment by nmoinvaz** [comment]
Do you have example source code that exhibits the problem? The change was made to allow `mz_zip_read_entry()` function to return the decrypted zip data (deflate data). If `password == NULL` then it doesn't decrypt the data, otherwise it does.

**Comment by vszakats** [comment]
The problem I'm hitting happens when _creating_ the raw (encrypted) entry in the target .zip (based on a raw entry of an encrypted .zip entry). In such scenario the [`if()`](https://github.com/nmoinvaz/minizip/blob/6781995724f41facba279152f15f16f680154582/mz_zip.c#L1138-L1139) block should _not_ be executed, otherwise it hits the `#ifdef HAVE_AES` or `#ifdef HAVE_PKCRYPT` protected codepaths, creating a duplicate encryption header (over the copied raw entry payload) and making the result invalid.

I have working source code that I used to reduce the issue to this single hunk, but it's written in a high level language (which is a slightly modified version of production code) over the minizip wrapper. It's possible to rewrite this in pure C, but it'd take quite some time.

When _reading_ the source entry, the code is using raw mode without passing a password and expecting to read the entry as-is.

Here's the relevant high-level source code:
```harbour
               ELSEIF hb_unzipFileOpen( fhndRef,, .T. /* lRaw */ ) == UNZ_OK .AND. ;
                      hb_zipFileCreate( fhnd, cFileO, tDate,, nAttrI, nAttrE, ;
                                        nMethod, nLevel, "" /* any non-null password to flag the entry as encrypted */,, ;
                                        cComment,, ;
                                        .T. /* lRaw */, ;
                                        lZip64 ) == ZIP_OK
```

**Comment by vszakats** [comment]
This expression, when called from `mz_zip_entry_write_open()` in _raw mode_:
```
    if ((err == MZ_OK) && (zip->file_info.flag & MZ_ZIP_FLAG_ENCRYPTED) &&.
        (zip->compression_method != MZ_COMPRESS_METHOD_RAW || password != NULL))
```

will have these two possible outcomes depending on `password` being `non-NULL` or `NULL`:

`non-NULL`
```
    if (true && true &&
        (false || true))
```
In this case the entry will be flagged as encrypted and extra encryptions headers (which are already present in the raw data) get added.

`NULL`
```
    if (true && false &&
        (false || false))
```
In this case the entry will _not_ be flagged as encrypted, but the raw (encrypted) data will be dumped without any extra headers.

---

Notice that `zip->file_info.flag & MZ_ZIP_FLAG_ENCRYPTED` and `password != NULL` always gets the same value, the caller cannot do anything to make them different, even though for the correct result, the first expression should be `true` (for an encrypted entry) and the second `false` (to not add extra encryption headers in raw mode).

TL;DR: Extra encryption headers should only be written if _not in raw mode_ and _writing an encrypted entry_. It means that for writing an entry, the `password != NULL` condition is unnecessary.

**Comment by nmoinvaz** [comment]
If you remove `zip->file_info.flag &= ~MZ_ZIP_FLAG_ENCRYPTED;` in `mz_zip_entry_write_open` will that would for you?

**Comment by vszakats** [comment]
I've tried it now, but it still results in invalid output, because in this case it missed a 11-byte segment (that turned out to be the AES header) from the PK headers (both local and central), so the result was 22 bytes shorter than the valid output.

The reason for that is that `file_info.aes_version` is only set to a non-zero value if `password != NULL` inside `zipOpenNewFileInZip5()`, and the required AES header is only written if this value is non-zero.

So with this additional modification below, it _does_ work, it also does break compatibility with minizip1 by relying on a passed encryption flag and behaving differently with the passed `password` value, which isn't the most ideal IMO:
```diff
--- a/mz_compat.c
+++ b/mz_compat.c
@@ -162,7 +162,7 @@ extern int ZEXPORT zipOpenNewFileInZip5(zipFile file, const char *filename, cons
     else
         file_info.zip64 = MZ_ZIP64_DISABLE;
 #ifdef HAVE_AES
-    if (password)
+    if (password || (raw && (file_info.flag & MZ_ZIP_FLAG_ENCRYPTED)))
         file_info.aes_version = MZ_AES_VERSION;
 #endif
```

**Comment by vszakats** [comment]
Back to breaking minizip1 compatibility: After reviewing it more closely, it seems that I most likely used minizip1 wrong, and it also does expect `password == NULL` and setting the encryption flag manually when writing a raw entry.

So it's seem to be fine go with this patch:
```
diff --git a/mz_compat.c b/mz_compat.c
index e96bdedc7b..52857e44ac 100644
--- a/mz_compat.c
+++ b/mz_compat.c
@@ -162,7 +162,7 @@ extern int ZEXPORT zipOpenNewFileInZip5(zipFile file, const char *filename, cons
     else
         file_info.zip64 = MZ_ZIP64_DISABLE;
 #ifdef HAVE_AES
-    if (password)
+    if (password || (raw && (file_info.flag & MZ_ZIP_FLAG_ENCRYPTED)))
         file_info.aes_version = MZ_AES_VERSION;
 #endif
 
diff --git a/mz_zip.c b/mz_zip.c
index 1f13ea0453..8d5f3759f2 100644
--- a/mz_zip.c
+++ b/mz_zip.c
@@ -1340,8 +1345,6 @@ extern int32_t mz_zip_entry_write_open(void *handle, const mz_zip_file *file_inf
 
     if (password != NULL)
         zip->file_info.flag |= MZ_ZIP_FLAG_ENCRYPTED;
-    else
-        zip->file_info.flag &= ~MZ_ZIP_FLAG_ENCRYPTED;
 
     mz_stream_get_prop_int64(zip->stream, MZ_STREAM_PROP_DISK_NUMBER, &disk_number);
     zip->file_info.disk_number = (uint32_t)disk_number;
```

**Comment by nmoinvaz** [comment]
Thanks! I have made the changes.

**Comment by vszakats** [comment]
Thank you!
